### PR TITLE
Implement support for async generator functions and for-await statements

### PR DESCRIPTION
**Pull request by zenparsing**
*This PR depends on https://github.com/babel/babylon/commit/b926e401c673a74e76a253e008d92f1e83024fdc.  Tests will fail until the next babylon release, but submitting early for feedback.*

This change implements the [async iteration proposal](https://github.com/tc39/proposal-async-iteration), currently at stage 2.  It includes the following features:

- Transforms async generator functions (`async function* g() { }`) to wrapped generator functions, similar to the current async-to-generator transform.
- Transforms `for-await` statements into for loops containing yield expressions.

Thanks!

### Implementation Notes

*Additional comments inline.*

#### Async Generator Functions

Async generator functions are transformed into regular generator functions, wrapped with a call to a new helper function (`asyncGenerator.wrap`).  Within the async generator function, await expressions are translated into yield expressions, where the argument is "boxed" using another helper function (`asyncGenerator.await`).  The wrapper function detects instances of the "boxed" values and knows that it should "pump" the generator with the awaited value instead of yielding to the consumer.

Rather than duplicate the async wrapping transform in `babel-helper-remap-async-to-generator`, I chose to refactor that helper to allow an optional await wrapping callee.

#### for-await

Since `for-await` statements can also appear inside of regular async functions and must take into account whether or not await expressions are "boxed", the `for-await` translation is done inside `babel-helper-remap-async-to-generator`.  Most of the `for-await` transform logic comes from the existing `for-of` transformer.

### CODE REVIEW

File: packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/object-method-with-arrows/expected.js [hunk 1]
```diff
@@ -16,9 +16,9 @@ class Class {
           var _this2 = this;
 
           this;
-          (function () {
-            _this2;
-          });
+          () => {
```
Comment (zenparsing) [inline packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/object-method-with-arrows/expected.js]: I looks to me like the current async-to-generator transform is unnecessarily transforming arrows that are nested within nested function declarations and expressions.

Is there some reason why arrows nested within a nested function would need to be translated?

**Comment by jprichardson** [comment]
Since this functionality was in Babel v5, any idea when this will make it into the current Babel?

**Comment by hzoo** [comment]
Made the babylon release https://github.com/babel/babylon/releases/tag/v6.8.0

**Comment by zenparsing** [comment]
Can someone help explain the failing "codecov/changes" check?  I can quite figure out what's going on there.

**Comment by hzoo** [comment]
I think there are some issues with codecov (recently), I wouldn't worry about it atm

**Comment by zenparsing** [comment]
OK.  It seems like we should have some eval tests, but I'm not sure if that's really possible given the async nature of this transform?  Are there any functional tests for async functions?

**Comment by hzoo** [comment]
I don't think there have been - only https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-async-to-generator/test/fixtures/regression/T6882/exec.js which was a test for hoisting http://phabricator.babeljs.io/T6882. Otherwise I would check out the existing exec.js tests

**Comment by jprichardson** [comment]
Anymore progress on this? This unlocks a lot of potential and code broke since it was in Babel v5 :(

**Comment by baslr** [comment]
Then, finally we can iterate nicely over a database query cursor

**Comment by jprichardson** [comment]
Is there anything I can help with here to get this integrated into a Babel release?

**Comment by zenparsing** [comment]
@jprichardson There a failing code coverage test which I can't make heads or tails of, and there really ought to be some sort of "exec" tests, but I haven't had time to sit down and figure out how to make that work for async operations.

**Comment by hzoo** [comment]
@zenparsing Don't worry about the coverage issue like I mentioned above

**Comment by wifiextender** [comment]
Just found out that Babel does **NOT** transpile `arr.includes` to ES5 `arr.indexOf`.

**Comment by loganfsmyth** [comment]
@wifiextender Correct, you load `babel-polyfill` for that. This has nothing to do with this thread though.

**Comment by tonyxiao** [comment]
According to `transform-regenerator` docs we already support async generators? https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-regenerator/README.md

**Comment by Kovensky** [comment]
@tonyxiao that is if you use the regenerator runtime. This seems to be for use with native generators.

**Comment by tonyxiao** [comment]
I see. @Kovensky got it thanks.

**Comment by alexeyraspopov** [comment]
@zenparsing, do you still have plans to finish this? It will be awesome!

**Comment by jprichardson** [comment]
Isn't this now possible with this https://github.com/babel/babel/tree/master/packages/babel-plugin-syntax-async-generators ?

**Comment by hzoo** [comment]
Yes but we need a code review and some more exec tests as mentioned above - unless we want users to be our testers 😛

**Comment by domenic** [comment]
FYI, we're hoping to push the async iteration proposal to stage 3 at the next TC39 meeting (in two weeks' time). I totally understand if the Babel maintainers don't have the time to take this pull request over the finish line, but it would be ideal if Babel supported such a stage 3 feature.

**Comment by hzoo** [comment]
I'l try to take a look this week or at least finish up the tests so we can get this used/tested

**Comment by hzoo** [comment]
@zenparsing can you fixup the merge conflicts? If I don't get around to the tests we'll just merge as is so we can get something out there like @domenic said

**Comment by hzoo** [comment]
Going to merge manually in another pr

**Comment by hzoo** [comment]
Made https://github.com/babel/babel/pull/4576

**Comment by hzoo** [comment]
Thanks @zenparsing, sorry for the super long delay

**Pull request closed without merging**